test/test_entrypoint.py

```diff
@@ -0,0 +1,136 @@
+#!/usr/bin/python3
+
+import unittest
+import logging
+import inspect
+import tempfile
+import os
+import sys
+import shutil
+
+import juju.entrypoint as ep
+import juju.charm
+
+from unittest.mock import patch
+
+from pathlib import Path
+
+
+logger = logging.getLogger()
+logger.level = logging.DEBUG
+stream_handler = logging.StreamHandler(sys.stdout)
+logger.addHandler(stream_handler)
+
+
+class TestEntrypoint(unittest.TestCase):
+
+    def setUp(self):
+        self.tmpdir = Path(tempfile.mkdtemp())
+        db_path_patch = patch('juju.entrypoint.DB_PATH',
+                              os.path.join(self.tmpdir, '.unit-state.db'))
+        db_path_patch.start()
+        self.addCleanup(db_path_patch.stop)
+
+    def tearDown(self):
+        shutil.rmtree(self.tmpdir)
+
+    def test_event_reemitted(self):
+        # note: class variables are modified because Charm instances
+        # are not preserved across simulated hook executions
+        class CharmConfigDeferrer(juju.charm.Charm):
+            defer = True
+            count_called = 0
+            count_deferred = 0
+
+            def __init__(self, framework):
+                super().__init__(framework)
+                self.framework.observe(self.on.config_changed, self)
+
+            def on_config_changed(self, event):
+                logger.debug(f'In {inspect.stack()[0].function}')
+                if self.defer:
+                    event.defer()
+                    type(self).defer = False
+                    type(self).count_deferred += 1
+                type(self).count_called += 1
+
+        with patch('juju.entrypoint.Charm', CharmConfigDeferrer):
```

Comment:
Again, we should be able to avoid replacing code like this in most cases, probably. Let's sync up on this and see what we can come up with.

Reply:
OK, sounds good - happy to replace it with a better way if we find one.

I provided an explanation around using a class called `Charm` and a module called `charm.py` in the comment above. Here I simply substitute the blank declaration from `charm.py` with the one under test.